src/main/java/com/linkedin/kafka/clients/producer/ExtensibleProducerRecord.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2017 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License").
+ * See License in the project root for license information.
+ */
+package com.linkedin.kafka.clients.producer;
+
+import com.linkedin.kafka.clients.utils.HeaderUtils;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+
+/**
+ * <p>
+ * This class adds the capability to have key-value pairs, headers, which are delivered to the consumer.  The
+ * consumer must be using {@link com.linkedin.kafka.clients.consumer.LiKafkaConsumer}.  If headers are present and the
+ * consumer is not using
+ * {@link com.linkedin.kafka.clients.consumer.LiKafkaConsumer} then the consumer records will not be deserializable.
+ * True null values can not be sent in the presence of any headers.
+ * </p>
+ * <p>{@link HeaderUtils} contains suggestions for how to partition the header key into
+ * intervals.  Header keys must be non-null, non-zero length strings that can be encoded into a utf-8 string.  Headers
+ * live in a separate space from the underlying Kafka protocol.
+ * </p>
+ * <p> A header can be added to a producer record by calling {@link #header(String, byte[])}.  This implies that the
```

Comment:
In the consumer record:
We want headers to be mutable because we want to have plugins at some point that support modification of headers.  Right now I need this in order to remove the large message header otherwise we could just leave the header there.  Since the large message support is not transparent to the end user I think it's fine to just leave the large message header.

In the producer record (which this comment specifically concerns):
We hide the data structure used to store the headers from the user so either we expose that we are using a Map or else we have some kind of HeaderBuilder that builds a Collection<Header> that is then passed to the producer constructor.

Since we have byte[] values, unless we do expensive defensive copying for each record, the record is still mutable.